lib/cxx-api/include/clingo/control.hh

```diff
@@ -349,6 +289,67 @@ class Control {
         ground(PartSpan{parts}, std::move(ctx));
     }
 
+    //! Ground the logic program with the given parameters.
+    //!
+    //! The given parts determine which program parts are grounded with which
+    //! parameters. If no parts are given, those specified by the parts
+    //! directive are grounded.
+    //!
+    //! A handler implementing the GroundEventHandler interface can be given to
+    //! handle grounding events. The handler is passed by value; use std::ref
+    //! if you want to manage its lifetime externally. The handler is optional
+    //! and can be set to nullptr to ignore events.
+    //!
+    //! @param parts the parts to ground the control object with
+    //! @param handler the context to use for grounding
+    template <Detail::UserData<GroundEventHandler> Handler = std::nullptr_t>
+    [[nodiscard]] auto start_ground(std::optional<PartSpan> parts = std::nullopt, Handler &&handler = nullptr) const
+        -> GroundHandle {
+        auto [default_parts, c_parts] = c_parts_(parts);
+        auto user_data = Detail::make_user_data_manager(std::forward<Handler>(handler));
+        using UserData = decltype(user_data);
+        clingo_ground_event_handler_t const *c_handler_ptr = nullptr;
+        if constexpr (!std::is_null_pointer_v<typename UserData::ValueType>) {
+            static constexpr auto c_handler = clingo_ground_event_handler_t{
+                [](char const *name, size_t name_size, size_t arguments_size, void *data, bool *result) -> bool {
+                    CLINGO_TRY {
+                        *result = UserData::cast(data)->callable(std::string_view{name, name_size}, arguments_size);
+                    }
+                    CLINGO_CATCH;
+                },
+                []([[maybe_unused]] clingo_lib_t *lib, clingo_location_t const *location, char const *name,
+                   size_t name_size, clingo_symbol_t const *arguments, size_t arguments_size, void *data,
+                   clingo_symbol_callback_t symbol_callback, void *symbol_callback_data) -> bool {
+                    CLINGO_TRY {
+                        Location loc{location};
+                        auto syms =
+                            UserData::cast(data)->call(loc, {name, name_size}, {cpp_cast(arguments), arguments_size});
+                        auto const *c_syms = c_cast(syms.data());
+                        return symbol_callback(c_syms, syms.size(), symbol_callback_data);
+                    }
+                    CLINGO_CATCH;
+                },
+                [](clingo_ground_result_t result, void *data) {
+                    UserData::cast(data)->finish(static_cast<GroundResult>(result));
+                },
+                [](void *data) { UserData::free(data); },
+            };
+            c_handler_ptr = &c_handler;
+        }
+        clingo_ground_handle_t *handle = nullptr;
+        Detail::handle_error(clingo_control_start_ground(ctl_.get(), c_parts.data(), c_parts.size(),
```

Comment:
Please consider using `c_handler_ptr` directly here. I.e.:
```cpp
Detail::handle_error(clingo_control_start_ground(ctl_.get(), c_parts.data(), c_parts.size(), c_handler_ptr,
                                                         user_data.release(), &handle));
```
It has the same effect but is more obvious and avoids confusion (as written the code would be incorrect if one would (naively) assume `user_data.release()` behaves like  `unique_ptr::release`).